ops/opset13/xor.go

```diff
@@ -0,0 +1,108 @@
+package opset13
+
+import (
+	"github.com/advancedclimatesystems/gonnx/onnx"
+	"github.com/advancedclimatesystems/gonnx/ops"
+	"gorgonia.org/tensor"
+)
+
+var (
+	MinXorInputs = 2
+	MaxXorInputs = 2
+)
+
+// Xor represents the ONNX xor operator.
+type Xor struct{}
+
+// newXor creates a new xor operator.
+func newXor() ops.Operator {
+	return &Xor{}
+}
+
+// Init initializes the xor operator.
+func (x *Xor) Init(_ []*onnx.AttributeProto) error {
+	return nil
+}
+
+// Apply applies the xor operator.
+func (x *Xor) Apply(inputs []tensor.Tensor) ([]tensor.Tensor, error) {
+	in1, in2, err := ops.MultidirectionalBroadcast(inputs[0], inputs[1])
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := xor(in1, in2)
+	if err != nil {
+		return nil, err
+	}
+
+	return []tensor.Tensor{out}, nil
+}
+
+// ValidateInputs validates the inputs that will be given to Apply for this operator.
+func (x *Xor) ValidateInputs(inputs []tensor.Tensor) ([]tensor.Tensor, error) {
+	return ops.ValidateInputs(x, inputs)
+}
+
+// GetMinInputs returns the minimum number of input tensors this operator expects.
+func (x *Xor) GetMinInputs() int {
+	return MinXorInputs
+}
+
+// GetMaxInputs returns the maximum number of input tensors this operator expects.
+func (x *Xor) GetMaxInputs() int {
+	return MaxXorInputs
+}
+
+// GetInputTypeConstraints returns a list. Every element represents a set of allowed tensor dtypes
+// for the corresponding input tensor.
+func (x *Xor) GetInputTypeConstraints() [][]tensor.Dtype {
+	return [][]tensor.Dtype{{tensor.Bool}, {tensor.Bool}}
+}
+
+// String implements the stringer interface, and can be used to format errors or messages.
+func (x *Xor) String() string {
+	return "xor operator"
+}
+
+func xor(A, B tensor.Tensor) (tensor.Tensor, error) {
```

Comment:
These operations look very similar. The only difference i can see is the actual operation on line 96. Maybe you could implement a more general `boolOp` func that takes the tensors handles the iteration stuff and also takes an `Op` func that takes two bools and returns a bool:

```go
type BooleanOperator func(a, b bool) bool 

func applyBooleanOperator(A, B tensor.Tensor, op BooleanOperator) (tensor.Tensor, error) {
Comment
	output := tensor.NewDense(tensor.Bool, A.Shape())
	output.Zero()

	iterator := A.Iterator()
	iterator.Reset()

	for !iterator.Done() {
		valA, err := A.At(iterator.Coord()...)
		if err != nil {
			return nil, err
		}

		boolA, ok := valA.(bool)
		if !ok {
			return nil, ops.ErrTypeAssert("bool", valA)
		}

		valB, err := B.At(iterator.Coord()...)
		if err != nil {
			return nil, err
		}

		boolB, ok := valB.(bool)
		if !ok {
			return nil, ops.ErrTypeAssert("bool", valB)
		}

		err = output.SetAt(op(boolA, boolB), iterator.Coord()...)
		if err != nil {
			return nil, err
		}

		_, err = iterator.Next()
		if err != nil {
			return nil, err
		}
	}

	return output, nil
}
``` 
Maybe you could go a bit further and add the broadcasting as wel this way all the operators would be a simple oneliner. And the complexity of the iteration is kept away behind a single function.